fe/fe-core/src/main/java/com/starrocks/qe/SetExecutor.java

```diff
@@ -89,8 +97,22 @@ private void setVariablesOfAllType(SetListItem var) throws DdlException {
      * @throws DdlException
      */
     public void execute() throws DdlException {
-        for (SetListItem var : stmt.getSetListItems()) {
-            setVariablesOfAllType(var);
+        HashMap<String, UserVariable> cloneUserVars = new HashMap<>();
+        boolean hasUserVar = stmt.getSetListItems().stream().anyMatch(var -> var instanceof UserVariable);
+        if (hasUserVar) {
+            cloneUserVars.putAll(ctx.getUserVariables());
+        }
+        try {
+            for (SetListItem var : stmt.getSetListItems()) {
+                setVariablesOfAllType(var);
+            }
+        } catch (Throwable e) {
+            //If the set sql contains more than one user variable,
+            //the atomicity of the modification of this set of variables must be ensured.
+            if (hasUserVar) {
+                ctx.setUserVariables(cloneUserVars);
```

Comment:
Would it be a better choice to put the user variable into a temporary container first and then put it in the last time?